remote_attestation_sessions/src/client.rs

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::SessionId;
+use anyhow::Context;
+use async_trait::async_trait;
+use oak_remote_attestation::handshaker::{
+    AttestationBehavior, ClientHandshaker, Encryptor, ServerIdentityVerifier,
+};
+
+#[async_trait]
+pub trait UnaryClient {
+    async fn message(&mut self, session_id: SessionId, body: Vec<u8>) -> anyhow::Result<Vec<u8>>;
+}
+
+/// gRPC Attestation Service client implementation.
+pub struct GenericAttestationClient<T: UnaryClient> {
+    session_id: SessionId,
+    encryptor: Encryptor,
+    client: T,
+}
+
+impl<T: UnaryClient> GenericAttestationClient<T> {
+    pub async fn create(
+        mut client: T,
+        expected_tee_measurement: &[u8],
+        server_verifier: ServerIdentityVerifier,
+    ) -> anyhow::Result<Self> {
+        let session_id: SessionId = rand::random();
+
+        let mut handshaker = ClientHandshaker::new(
+            AttestationBehavior::create_peer_attestation(expected_tee_measurement),
+            server_verifier,
+        );
+        let client_hello = handshaker
+            .create_client_hello()
+            .context("Couldn't create client hello message")?;
+
+        let mut response = client
+            .message(session_id, client_hello)
+            .await
+            .context("Couldn't message client hello message")?;
+
+        while !handshaker.is_completed() {
+            let request = handshaker
+                .next_step(&response)
+                .context("Couldn't process handshake message")?;
+
+            if let Some(request) = request {
```

Comment:
What if not? Also, is it worth adding some debug logging throughout, or do you think just returning the errors should be enough in practice?

Reply:
If not the handshake is done. This goes back to your previous comment, there's some implicit coupling between this and `is_completed()`. Back in the day I listed this in the refactoring ticket: https://github.com/project-oak/oak/issues/2608

Reply:
Also pre-existent, this PR just moves the method but does not really modify it